### PR TITLE
Google Workspace File Parsers

### PR DESCRIPTION
**Pull request by vakohl**
Required items, please complete
   
   Change(s):
   - Added new ASIM File Parsers for Google Workspace

   Reason for Change(s):
   - New parsers created

   Version Updated:
   - 0.1.0

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Will take care

### CODE REVIEW

File: Parsers/ASimFileEvent/test/GoogleWorkspace_vimFileEvent_SchemaTest.csv [hunk 1]
```diff
@@ -0,0 +1,92 @@
+﻿Result
+"(0) Error: Missing mandatory field [Type]"
```
Comment (anki-narravula) [inline Parsers/ASimFileEvent/test/GoogleWorkspace_vimFileEvent_SchemaTest.csv:2]: Why this error coming ? Need test case correction ?

**Review by anki-narravula — CHANGES_REQUESTED** [review]

File: Parsers/ASimFileEvent/Parsers/ASimFileEventGoogleWorkspace.yaml [hunk 2]
```diff
@@ -0,0 +1,286 @@
+Parser:
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimFileEvent/Parsers/ASimFileEventGoogleWorkspace.yaml:1]: thanks @anki-narravula. updated now.

File: Parsers/ASimFileEvent/Parsers/vimFileEventGoogleWorkspace.yaml [hunk 3]
```diff
@@ -0,0 +1,343 @@
+Parser:
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimFileEvent/Parsers/vimFileEventGoogleWorkspace.yaml:1]: Yes, has been updated now. thankyou

File: Parsers/ASimFileEvent/Parsers/ASimFileEventGoogleWorkspace.yaml [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                  "FolderMoved",
+                  TargetFileMimeType == "folder" and event_name_s == "untrash",
+                  "FolderCreated",
+                  EventType
+              ),
+          EventSubType = case(
+                    TargetFileMimeType == "folder" and event_name_s == "create",
+                    "",
+                    TargetFileMimeType == "folder" and event_name_s == "trash",
+                    "",
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimFileEvent/Parsers/ASimFileEventGoogleWorkspace.yaml:150]: @anki-narravula 
EventSubType is not mandatory, so we can keep as null if condition not met.

File: Parsers/ASimFileEvent/Parsers/ASimFileEventGoogleWorkspace.yaml [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                    event_name_s == 'approval_due_time_change',
+                    strcat(ActorUsername, " requested a due time change on an approval"),
+                    event_name_s == 'approval_requested',
+                    strcat(ActorUsername, " requested approval on an item"),
+                    event_name_s == 'approval_reviewer_change',
+                    strcat(ActorUsername, " requested a reviewer change on an approval"),
+                    event_name_s == 'approval_reviewer_responded',
+                    strcat(ActorUsername, " reviewed an approval on an item"),
+                    event_name_s == 'create_comment',
+                    strcat(ActorUsername, " created a comment"),
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimFileEvent/Parsers/ASimFileEventGoogleWorkspace.yaml:207]: @anki-narravula I thought about it, the problem was, we can't keep dynamic values in table.

File: Parsers/ASimFileEvent/Parsers/vimFileEventGoogleWorkspace.yaml [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+                        "Copy_Type",
+                        copy_type_s,
+                        "Shared_Drive_Id",
+                        shared_drive_id_s,
+                        "Destination_Folder_Id",
+                        destination_folder_id_s,
+                        "Source_Folder_Id",
+                        source_folder_id_s
+                    )
+      | where ((array_length(srcfilepath_has_any) == 0) or (SrcFilePath has_any (srcfilepath_has_any)))
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimFileEvent/Parsers/vimFileEventGoogleWorkspace.yaml:298]: @anki-narravula  I felt it was appropriate as SrcFilePath value is applicable only when Event Type is ('rename', 'copy', 'source_copy'). Defined at line 173.

File: Parsers/ASimFileEvent/test/GoogleWorkspace_ASimFileEvent_DataTest.csv [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 1372 records (8.94%) for field [IpAddr] of type [IP Address]: [""-""] (Schema:FileEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 1372 records (8.94%) for field [SrcIpAddr] of type [IP Address]: [""-""] (Schema:FileEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 1372 records (8.94%) for field [Src] of type [IP Address]: [""-""] (Schema:FileEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 15344 records (100.0%) for field [ActorUserIdType] of type [Enumerated]: [""GWorkspaceProfileID""] (Schema:FileEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 15344 records (100.0%) for field [TargetFilePathType] of type [Enumerated]: [""FilenameOnly""] (Schema:FileEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 476 records (3.1%) for field [SrcFilePathType] of type [Enumerated]: [""FilenameOnly""] (Schema:FileEvent)"
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimFileEvent/test/GoogleWorkspace_ASimFileEvent_DataTest.csv:7]: @anki-narravula you meant the error coming for ActorUserIdType, TargetFilePathType, SrcFilePathType fields?
I'm already updating Tester.csv with these supported values but not sure why it still complains.

File: Parsers/ASimFileEvent/test/GoogleWorkspace_vimFileEvent_SchemaTest.csv (hunk 1 above)
Comment (vakohl) [reply Parsers/ASimFileEvent/test/GoogleWorkspace_vimFileEvent_SchemaTest.csv:2]: @anki-narravula 
strange, I re-run test today. It did not appear. I think we are good.
image

Comment (vakohl) [reply Parsers/ASimFileEvent/test/GoogleWorkspace_vimFileEvent_SchemaTest.csv:2]: also, updated results in PR

**Comment by vakohl** [comment]
Thanks @anki-narravula for the review. I've looked at all comments. Please validate.

**Review by anki-narravula — DISMISSED** [review]

**Review by anki-narravula — APPROVED** [review]

**Pull request merged**